examples/pytorch/__temporary__/dglnew/graph/other_feature.py

```diff
@@ -0,0 +1,26 @@
+
+# A GraphStorage class where ndata and edata can be any FeatureStorage but
+# otherwise the same as the wrapped DGLGraph.
+class OtherFeatureGraphStorage(object):
+    def __init__(self, g, ndata=None, edata=None):
+        self.g = g
+        self._ndata = ndata or {}
+        self._edata = edata or {}
+
+    @property
+    def ndata(self):
+        return self._ndata
+
+    @property
+    def edata(self):
+        return self._edata
+
+    def __getattr__(self, key):
+        # I wrote it in this way because I'm too lazy to write "def sample_neighbors"
+        # or stuff like that.
+        if key in ['ntypes', 'etypes', 'canonical_etypes', 'sample_neighbors',
+                   'subgraph', 'edge_subgraph', 'find_edges', 'num_nodes']:
```

Comment:
One issue with this design is if I come up with a custom sampler (e.g., TimeSensitiveEdgeDataLoader), which uses the custom method (e.g., `sample_historical_neighbors()`) which I've added to my GraphStorage class (probably a child of DGLGraph, but perhaps not), the sampler would no longer be able to invoke the custom sampling function since it's not in this list.

Could we decouple our GraphStorage and FeatureStorage, such both get passed into a dataloader (for backward compatibility allow passing in `feature_storage=None` if the GraphStorage object is also a FeatureStore-like object)?

Reply:
> One issue with this design is if I come up with a custom sampler (e.g., TimeSensitiveEdgeDataLoader), which uses the custom method (e.g., sample_historical_neighbors())
Do you have code online?  `TimeSensitiveEdgeDataLoader` sounds just like what TGAT/TGN needed.  I would try making it work with DGLGraphs if necessary.

> the sampler would no longer be able to invoke the custom sampling function since it's not in this list.
This `__getattr__` implementation is just an alternative to:
```
@property
def ntypes(self):
    return self.g.ntypes

@property
def etypes(self):
    return self.g.etypes
...
```
It's **not** meant to be a spec for all other graph storages like "all graph storages must have and only have this list of methods".  If you have a new sampling primitive for your GraphStorage, the sampler could still call your custom sampling function via duck-typing.  Then it will be DGL's job to catch up with your new primitive.

> Could we decouple our GraphStorage and FeatureStorage, such both get passed into a dataloader (for backward compatibility allow passing in feature_storage=None if the GraphStorage object is also a FeatureStore-like object)?
What is the use case?  I didn't quite understand why a GraphStorage could also be a FeatureStorage.

Reply:
Is `OtherFeatureGraphStorage` your testing code or something that will be checked into the main repo?